extensions/flyway/deployment/src/main/java/io/quarkus/flyway/ScannerTransformer.java

```diff
@@ -0,0 +1,121 @@
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.MethodVisitor;
+
+import io.quarkus.flyway.runtime.QuarkusPathLocationScanner;
+import io.quarkus.gizmo.Gizmo;
+
+/**
+ * Transforms {@link Scanner} in a way to take advantage of our build time knowledge
+ * This should be removed completely if https://github.com/flyway/flyway/issues/2822
+ * is implemented
+ */
+class ScannerTransformer implements BiFunction<String, ClassVisitor, ClassVisitor> {
+
+    static final String FLYWAY_SCANNER_CLASS_NAME = Scanner.class.getName();
+    private static final String FLYWAY_SCANNER_INTERNAL_CLASS_NAME = FLYWAY_SCANNER_CLASS_NAME.replace('.', '/');
+
+    private static final String FLYWAY_RESOURCE_AND_CLASS_SCANNER_CLASS_NAME = ResourceAndClassScanner.class.getName();
+    private static final String FLYWAY_RESOURCE_AND_CLASS_SCANNER_INTERNAL_CLASS_NAME = FLYWAY_RESOURCE_AND_CLASS_SCANNER_CLASS_NAME
+            .replace('.', '/');
+
+    private static final String QUARKUS_RESOURCE_AND_CLASS_SCANNER_CLASS_NAME = QuarkusPathLocationScanner.class.getName();
+    private static final String QUARKUS_RESOURCE_AND_CLASS_SCANNER_INTERNAL_CLASS_NAME = QUARKUS_RESOURCE_AND_CLASS_SCANNER_CLASS_NAME
+            .replace('.', '/');
+
+    private static final String CTOR_METHOD_NAME = "<init>";
+
+    @Override
+    public ClassVisitor apply(String s, ClassVisitor cv) {
+        return new ScannerVisitor(cv);
+    }
+
+    private static final class ScannerVisitor extends ClassVisitor {
+
+        public ScannerVisitor(ClassVisitor cv) {
+            super(Gizmo.ASM_API_VERSION, cv);
+        }
+
+        @Override
+        public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+            MethodVisitor mv = super.visitMethod(access, name, descriptor, signature, exceptions);
+            if (name.equals(CTOR_METHOD_NAME)) {
+                return new ConstructorTransformer(mv);
+            }
+            return mv;
+        }
+
+        /**
+         * Replaces the constructor of the {@link Scanner} with:
+         *
+         * <pre>
+         * public ScannerSubstitutions(Class<?> implementedInterface, Collection<Location> locations, ClassLoader classLoader,
+         *         Charset encoding, ResourceNameCache resourceNameCache, LocationScannerCache locationScannerCache) {
+         *     ResourceAndClassScanner quarkusScanner = new QuarkusPathLocationScanner(locations);
+         *     resources.addAll(quarkusScanner.scanForResources());
+         *     classes.addAll(quarkusScanner.scanForClasses());
+         * }
+         * </pre>
+         */
+        private static class ConstructorTransformer extends MethodVisitor {
+
+            public ConstructorTransformer(MethodVisitor mv) {
+                super(Gizmo.ASM_API_VERSION, mv);
+            }
+
+            @Override
+            public void visitCode() {
+                super.visitVarInsn(ALOAD, 0);
```

Comment:
I really should look at a 'Gizmo for transformers'...

Reply:
Yeah, that would make things a lot easier :)